File: Data Structures & Algorithms/minimum-window-with-characters/submission-0.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if(t.empty() || s.empty()){
            return "";
        }
        unordered_map<char, int> countWeHave;
        unordered_map<char, int> countWeNeed;

        for(int i=0;i<t.size();i++) {
            countWeNeed[t[i]]++;
        }
        int have = 0, need = countWeNeed.size();
        int res = INT_MAX;
        string substr = "";
        int subStart = 0, subEnd = 0;
        int l = 0;
        for(int r=0;r<s.size();r++) {
            if(countWeNeed.find(s[r]) != countWeNeed.end()){
                countWeHave[s[r]]++;
                if(countWeHave[s[r]] == countWeNeed[s[r]]) {
                    have += 1;
                }
            }
            
            while (have == need) {
                if((r - l + 1) < res) {
                    res = r - l + 1;
                    subStart = l;
                    subEnd = r;
                }
                if(countWeNeed.find(s[l]) != countWeNeed.end()){
                    countWeHave[s[l]]--;
                    if(countWeHave[s[l]] < countWeNeed[s[l]]) {
                        have -= 1;
                    } 
                }
                l++;
            }
        }
        return res == INT_MAX ? "" : s.substr(subStart, subEnd - subStart + 1);
     }
};

File: Data Structures & Algorithms/minimum-window-with-characters/submission-0.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(s) < len(t):
            return ""
        countT = defaultdict(int)
        for char in t:
            countT[char] += 1

        res = ""
        required = len(countT)
        current = 0
        countS = defaultdict(int)

        n = len(s)
        l, r = 0, 0
        while r < n:
            c = s[r]
            countS[c] += 1
            if c in countT and countS[c] == countT[c]:
                current += 1

            while l<= r and required == current:
                if res == "" or (r-l+1) < len(res):
                    res = s[l:r+1]
                countS[s[l]] -= 1
                if s[l] in countT and countS[s[l]] < countT[s[l]]:
                    current -= 1
                l += 1
            r += 1
        return res
        
